Subject: [PATCH] feat(!webhooks, lib): add webhook library (#109)

* feat(webhooks, lib): re-initialize after rebase origin

* refactor(webhooks): provide webhook entity in interceptors

* refactor: WebHookManager

* refactor: returns IQueryable for webhook subscriptions

* doc: update readme

* chore: revert subscription list to enumerable

---------

Co-authored-by: Maxim Vasin <maxim.vasin@mcc-soft.de>
---
Index: webapi/src/MccSoft.TemplateApp.App/Setup/SetupWebhooks.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/webapi/src/MccSoft.TemplateApp.App/Setup/SetupWebhooks.cs b/webapi/src/MccSoft.TemplateApp.App/Setup/SetupWebhooks.cs
new file mode 100644
--- /dev/null	(revision 0c7e929def77bf5a207871815ed49e62d8e8c1ee)
+++ b/webapi/src/MccSoft.TemplateApp.App/Setup/SetupWebhooks.cs	(revision 0c7e929def77bf5a207871815ed49e62d8e8c1ee)
@@ -0,0 +1,27 @@
+using MccSoft.TemplateApp.Domain.WebHook;
+using MccSoft.WebHooks;
+
+namespace MccSoft.TemplateApp.App.Setup;
+
+public partial class SetupWebhooks
+{
+    public static void AddWebhooks(WebApplicationBuilder builder)
+    {
+        builder.Services.AddWebHooks<TemplateWebHookSubscription>(optionsBuilder =>
+        {
+            // Define sequence of reties with delay in minutes.
+            // By default - it will retry once in a hour (60 minutes).
+            optionsBuilder.HangfireDelayInMinutes = [60, 120, 120];
+
+            // If you'd like to modify this class, consider adding your custom code in the SetupSwagger.partial.cs
+            // This will make it easier to pull changes from Template when Template is updated
+            // (actually this file will be overwritten by a file from template, which will make your changes disappear)
+            AddProjectSpecifics(builder, optionsBuilder);
+        });
+    }
+
+    static partial void AddProjectSpecifics(
+        WebApplicationBuilder builder,
+        WebHookOptionBuilder<TemplateWebHookSubscription> optionsBuilder
+    );
+}
Index: webapi/src/MccSoft.TemplateApp.App/Setup/SetupWebhooks.partial.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/webapi/src/MccSoft.TemplateApp.App/Setup/SetupWebhooks.partial.cs b/webapi/src/MccSoft.TemplateApp.App/Setup/SetupWebhooks.partial.cs
new file mode 100644
--- /dev/null	(revision 0c7e929def77bf5a207871815ed49e62d8e8c1ee)
+++ b/webapi/src/MccSoft.TemplateApp.App/Setup/SetupWebhooks.partial.cs	(revision 0c7e929def77bf5a207871815ed49e62d8e8c1ee)
@@ -0,0 +1,52 @@
+using MccSoft.TemplateApp.Domain.WebHook;
+using MccSoft.WebHooks;
+using Polly;
+
+namespace MccSoft.TemplateApp.App.Setup;
+
+public partial class SetupWebhooks
+{
+    static partial void AddProjectSpecifics(
+        WebApplicationBuilder builder,
+        WebHookOptionBuilder<TemplateWebHookSubscription> optionsBuilder
+    )
+    {
+        ConfigureResilienceOptions(optionsBuilder);
+        ConfigureInterceptors(builder, optionsBuilder);
+    }
+
+    private static void ConfigureResilienceOptions(
+        WebHookOptionBuilder<TemplateWebHookSubscription> optionsBuilder
+    )
+    {
+        optionsBuilder.ResilienceOptions.Delay = TimeSpan.FromSeconds(2);
+        optionsBuilder.ResilienceOptions.BackoffType = DelayBackoffType.Exponential;
+        optionsBuilder.ResilienceOptions.UseJitter = true;
+        optionsBuilder.ResilienceOptions.MaxRetryAttempts = 5;
+        optionsBuilder.ResilienceOptions.Timeout = TimeSpan.FromSeconds(30);
+    }
+
+    private static void ConfigureInterceptors(
+        WebApplicationBuilder builder,
+        WebHookOptionBuilder<TemplateWebHookSubscription> optionsBuilder
+    )
+    {
+        using ServiceProvider serviceProvider = builder.Services.BuildServiceProvider();
+        var logger = serviceProvider.GetRequiredService<ILogger<SetupWebhooks>>();
+        optionsBuilder.WebHookInterceptors.BeforeExecution = (webHook) =>
+        {
+            logger.LogInformation("Start delivering Webhook with ID: {WebHookId}", webHook?.Id);
+        };
+        optionsBuilder.WebHookInterceptors.AfterAllAttemptsFailed = (webHookId) =>
+        {
+            logger.LogInformation($"Webhook with ID: {webHookId} failed");
+        };
+        optionsBuilder.WebHookInterceptors.ExecutionSucceeded = (webHook) =>
+        {
+            logger.LogInformation(
+                "Webhook with ID: Webhook with ID: {WebHookId} delivered",
+                webHook?.Id
+            );
+        };
+    }
+}
Index: webapi/src/MccSoft.TemplateApp.Domain/WebHook/TemplateWebHookSubscription.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/webapi/src/MccSoft.TemplateApp.Domain/WebHook/TemplateWebHookSubscription.cs b/webapi/src/MccSoft.TemplateApp.Domain/WebHook/TemplateWebHookSubscription.cs
new file mode 100644
--- /dev/null	(revision 0c7e929def77bf5a207871815ed49e62d8e8c1ee)
+++ b/webapi/src/MccSoft.TemplateApp.Domain/WebHook/TemplateWebHookSubscription.cs	(revision 0c7e929def77bf5a207871815ed49e62d8e8c1ee)
@@ -0,0 +1,58 @@
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using MccSoft.DomainHelpers;
+using MccSoft.WebHooks.Domain;
+
+namespace MccSoft.TemplateApp.Domain.WebHook;
+
+/// <summary>
+/// This is an example of overriden webhook (to add Tenant for example. You can use base entity from lib)
+/// </summary>
+public class TemplateWebHookSubscription : WebHookSubscription, ITenantEntity
+{
+    #region ITenantEntity
+
+    public int TenantId { get; private set; }
+
+    private Tenant? _tenant;
+
+    public Tenant Tenant
+    {
+        get => _tenant.ThrowIfNotIncluded();
+        private set => _tenant = value;
+    }
+
+    public void SetTenantIdUnsafe(int tenantId)
+    {
+        TenantId = tenantId;
+    }
+
+    #endregion
+
+    /// <summary>
+    /// Needed for Entity Framework, keep empty.
+    /// </summary>
+    public TemplateWebHookSubscription() { }
+
+    public TemplateWebHookSubscription(
+        string name,
+        string url,
+        string eventType,
+        HttpMethod? method = null,
+        Dictionary<string, string>? headers = null
+    )
+        : base(name, url, eventType, (method ?? HttpMethod.Post).Method, headers) { }
+
+    #region Specification
+
+    /// <summary>
+    /// Creates a specification that is satisfied by a WebhookSubscription having the specified id.
+    /// </summary>
+    /// <param name="id">Subscription id.</param>
+    /// <returns>The created specification.</returns>
+    public static Specification<TemplateWebHookSubscription> HasId(Guid id) =>
+        new(nameof(HasId), x => x.Id == id, id);
+
+    #endregion
+}